src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,468 @@
+
+#include "libtorrent/aux_/escape_string.hpp"
+#include "libtorrent/aux_/session_settings.hpp"
+#include "libtorrent/aux_/websocket_tracker_connection.hpp"
+#include "libtorrent/utf8.hpp"
+#include "libtorrent/io.hpp"
+#include "libtorrent/ip_filter.hpp"
+#include "libtorrent/socket.hpp"
+#include "libtorrent/socket_io.hpp"
+#include "libtorrent/span.hpp"
+#include "libtorrent/tracker_manager.hpp"
+
+#include "libtorrent/aux_/disable_warnings_push.hpp"
+#include "boost/json.hpp"
+#include "libtorrent/aux_/disable_warnings_pop.hpp"
+
+#include <boost/system/system_error.hpp>
+
+#include <algorithm>
+#include <cctype>
+#include <cstdio> // for snprintf
+#include <exception>
+#include <functional>
+#include <list>
+#include <locale>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace libtorrent {
+namespace aux {
+
+using namespace std::placeholders;
+namespace errc = boost::system::errc;
+namespace error = boost::asio::error;
+namespace json = boost::json;
+
+namespace {
+
+// Overload for JSON strings
+std::string utf8_latin1(json::string_view sv)
+{
+	return libtorrent::utf8_latin1(std::string_view{sv.data(), sv.size()});
+}
+
+}
+
+websocket_tracker_connection::websocket_tracker_connection(io_context& ios
+		, tracker_manager& man
+		, tracker_request const& req
+		, std::weak_ptr<request_callback> cb)
+	: tracker_connection(man, req, ios, cb)
+	  , m_io_context(ios)
+	  , m_ssl_context(ssl::context::tlsv12_client)
+{
+	queue_request(req, cb);
+}
+
+void websocket_tracker_connection::start()
+{
+	if (is_started())
+		return;
+
+	auto const& settings = m_man.settings();
+	auto const& req = tracker_req();
+	m_websocket = std::make_shared<aux::websocket_stream>(m_io_context, m_man.host_resolver(), &m_ssl_context);
+
+	// in anonymous mode we omit the user agent to mitigate fingerprinting of
+    // the client. Private torrents is an exception because some private
+    // trackers may require the user agent
+    std::string const user_agent = settings.get_bool(settings_pack::anonymous_mode)
+			&& !req.private_torrent ? "" : settings.get_str(settings_pack::user_agent);
+	m_websocket->set_user_agent(user_agent);
+
+#ifndef TORRENT_DISABLE_LOGGING
+	if (auto cb = requester())
+		cb->debug_log("*** WEBSOCKET_TRACKER_CONNECT [ url: %s ]", req.url.c_str());
+#endif
+
+	ADD_OUTSTANDING_ASYNC("websocket_tracker_connection::on_connect");
+	m_websocket->async_connect(req.url, std::bind(&websocket_tracker_connection::on_connect
+				, this, weak_from_this(), _1));
+}
+
+void websocket_tracker_connection::close()
+{
+	if (m_websocket)
+	{
+		m_websocket->close();
+		m_websocket.reset();
+	}
+
+	error_code const ec{error::operation_aborted};
+	while (!m_pending.empty())
+	{
+		auto [msg, callback] = std::move(m_pending.front());
+		TORRENT_UNUSED(msg);
+		m_pending.pop();
+		if (auto cb = callback.lock())
+			cb->tracker_request_error(
+					tracker_req()
+                    , ec
+                    , operation_t::sock_read
```

Comment:
I think it would help debugging if you passed in the true operation that caused the cancellation and passed it on here. Otherwise you should probably use `operation_t::unknown`